libs/elasticsearch-core/src/main/java/org/elasticsearch/bootstrap/JarHell.java

```diff
@@ -0,0 +1,289 @@
+        String javaHome = System.getProperty("java.home");
+        logger.debug("java.home: {}", javaHome);
+        final Map<String,Path> clazzes = new HashMap<>(32768);
+        Set<Path> seenJars = new HashSet<>();
+        for (final URL url : urls) {
+            final Path path = PathUtils.get(url.toURI());
+            // exclude system resources
+            if (path.startsWith(javaHome)) {
+                logger.debug("excluding system resource: {}", path);
+                continue;
+            }
+            if (path.toString().endsWith(".jar")) {
+                if (!seenJars.add(path)) {
+                    throw new IllegalStateException("jar hell!" + System.lineSeparator() +
+                                                    "duplicate jar on classpath: " + path);
+                }
+                logger.debug("examining jar: {}", path);
+                try (JarFile file = new JarFile(path.toString())) {
+                    Manifest manifest = file.getManifest();
+                    if (manifest != null) {
+                        checkManifest(manifest, path);
+                    }
+                    // inspect entries
+                    Enumeration<JarEntry> elements = file.entries();
+                    while (elements.hasMoreElements()) {
+                        String entry = elements.nextElement().getName();
+                        if (entry.endsWith(".class")) {
+                            // for jar format, the separator is defined as /
+                            entry = entry.replace('/', '.').substring(0, entry.length() - 6);
+                            checkClass(clazzes, entry, path);
+                        }
+                    }
+                }
+            } else {
+                logger.debug("examining directory: {}", path);
+                // case for tests: where we have class files in the classpath
+                final Path root = PathUtils.get(url.toURI());
+                final String sep = root.getFileSystem().getSeparator();
+                Files.walkFileTree(root, new SimpleFileVisitor<Path>() {
+                    @Override
+                    public FileVisitResult visitFile(Path file, BasicFileAttributes attrs) throws IOException {
+                        String entry = root.relativize(file).toString();
+                        if (entry.endsWith(".class")) {
+                            // normalize with the os separator, remove '.class'
+                            entry = entry.replace(sep, ".").substring(0,  entry.length() - ".class".length());
+                            checkClass(clazzes, entry, path);
+                        }
+                        return super.visitFile(file, attrs);
+                    }
+                });
+            }
+        }
+    }
+
+    /** inspect manifest for sure incompatibilities */
+    private static void checkManifest(Manifest manifest, Path jar) {
```

Comment:
This looks like pure paranoia to me. I think we can keep this though as a bootstrap check. Then we do not need to do the version dance.